### PR TITLE
executor: push down tikv region status

### PR DESCRIPTION
**Pull request by qidi1**
### What problem does this PR solve?
The existing implementation of TIKV_REGION_STATUS  fetches all tables' regions infomation, which results in the same overhead of fetching a small table's regions infomation as a large one. To avoid this, some of the predicates can be pushed down to reduce unnecessary data transfer overhead.

Issue Number: #28330 

Problem Summary:

### What is changed and how it works?
Associated PR:
*  https://github.com/pingcap/tidb/pull/28211

Proposal: 
* https://github.com/pingcap/tidb/pull/28331

What's Changed:
 push down dbName,tableName,indexName,tableId,indexId for tikv region status

How it Works:

### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
'None' 
```

### CODE REVIEW

File: executor/memtable_reader.go [hunk 1]
```diff
@@ -698,3 +700,507 @@ func (e *clusterLogRetriever) close() error {
...
+func (e *hotRegionsHistoryRetriver) close() error {
+	if e.cancel != nil {
+		e.cancel()
+	}
+	return nil
+}
+
+func (e *hotRegionsHistoryRetriver) getRuntimeStats() execdetails.RuntimeStats {
+	return nil
+}
```
Comment (IcePigZDB) [inline executor/memtable_reader.go:1052]: it's work of previous pr, it should be add by `update branch` after previous pr merged.

File: executor/memtable_reader_test.go [hunk 2]
```diff
@@ -941,3 +944,706 @@ func (s *testMemTableReaderSuite) TestTiDBClusterLogError(c *C) {
...
+		warnings := tk.Se.GetSessionVars().StmtCtx.GetWarnings()
+		c.Assert(len(warnings), Equals, 0, Commentf("unexpected warnigns: %+v", warnings))
+		var expected []string
+		for _, row := range cas.expected {
+			expectedRow := row
+			expected = append(expected, strings.Join(expectedRow, " "))
+		}
+		result.Check(testkit.Rows(expected...))
+	}
+}
```
Comment (IcePigZDB) [inline executor/memtable_reader_test.go:1365]: ditto

File: executor/memtable_reader_test.go [hunk 3]
```diff
@@ -941,3 +944,706 @@ func (s *testMemTableReaderSuite) TestTiDBClusterLogError(c *C) {
...
+	c.Assert(rs.Close(), IsNil)
+
+	// Test without end time error.
+	rs, err = tk.Exec("select * from information_schema.tidb_hot_regions_history where update_time>='2019/08/26 06:18:13.011'")
+	c.Assert(err, IsNil)
+	_, err = session.ResultSetToStringSlice(context.Background(), tk.Se, rs)
+	c.Assert(err, NotNil)
+	c.Assert(err.Error(), Equals, "denied to scan hot regions, please specified the end time, such as `update_time < '2020-01-01 00:00:00'`")
+	c.Assert(rs.Close(), IsNil)
+}
```
Comment (IcePigZDB) [inline executor/memtable_reader_test.go:1388]: ditto

File: go.mod [hunk 4]
```diff
@@ -1,86 +1,102 @@
 module github.com/pingcap/tidb
```
Comment (IcePigZDB) [inline go.mod:1]: check is it necessary to include change in go.mod?

File: go.sum [hunk 5]
```diff
@@ -949,6 +949,8 @@ honnef.co/go/tools v0.0.1-2019.2.3/go.mod h1:a3bituU0lyd329TUQxRnasdCoJDkEUEAqEt
 honnef.co/go/tools v0.0.1-2020.1.3/go.mod h1:X/FiERA/W4tHapMX5mGpAtMSVEeEUOyHaw9vFzvIQ3k=
```
Comment (IcePigZDB) [inline go.sum:949]: ditto

File: planner/core/memtable_predicate_extractor.go [hunk 6]
```diff
@@ -1174,3 +1174,84 @@ func (e *TiFlashSystemTableExtractor) explainInfo(p *PhysicalMemTable) string {
...
+// Extract implements the MemTablePredicateExtractor Extract interface
+func (e *TikvRegionStatusExtractor) Extract(_ sessionctx.Context,
+	schema *expression.Schema,
+	names []*types.FieldName,
+	predicates []expression.Expression,
+) []expression.Expression {
+	remained, dbNameSkipRequest, dbNames := e.extractCol(schema, names, predicates, "db_name", true)
+	remained, tableNameSkipRequest, tableNames := e.extractCol(schema, names, remained, "table_name", true)
+	remained, indexNameSkipRequest, indexNames := e.extractCol(schema, names, remained, "index_name", true)
+	remained, tableIDSkipRequest, tableIDs := e.extractCol(schema, names, remained, "table_id", false)
```
Comment (IcePigZDB) [inline planner/core/memtable_predicate_extractor.go:1213]: how about split 5 fields into two group:
* `db_name`, `table_name`, `index_name` 
* `table_id`, `index_id`
and check SkipRequest twice.

File: planner/core/memtable_predicate_extractor_test.go [hunk 7]
```diff
@@ -1046,3 +1046,462 @@ func (s *extractorSuite) TestInspectionRuleTableExtractor(c *C) {
 		c.Assert(clusterConfigExtractor.SkipRequest, Equals, ca.skip, Commentf("SQL: %v", ca.sql))
 	}
 }
+
+func (s *extractorSuite) TestTiDBHotRegionsHistoryTableExtractor(c *C) {
```
Comment (IcePigZDB) [inline planner/core/memtable_predicate_extractor_test.go:1050]: ditto

File: planner/core/memtable_predicate_extractor_test.go [hunk 8]
```diff
@@ -1046,3 +1046,462 @@ func (s *extractorSuite) TestInspectionRuleTableExtractor(c *C) {
...
+}
+
+func (s *extractorSuite) TestTikvRegionStatusTableExtractor(c *C) {
+	var cases = []struct {
+		sql                             string
+		skipRequest                     bool
+		DBNames, TableNames, IndexNames set.StringSet
+		TableIDs, IndexIDs              set.Int64Set
+	}{
+		// Test full data, it will not call Extract() and executor(retriver) will panic and remind user to add conditions to save network IO.
```
Comment (IcePigZDB) [inline planner/core/memtable_predicate_extractor_test.go:1342]: the comment is not right, there is no fast error return here.

File: planner/core/util.go [hunk 9]
```diff
@@ -294,6 +294,32 @@ func extractStringFromStringSet(set set.StringSet) string {
...
+	l := make([]string, 0, len(set))
+	for k := range set {
+		l = append(l, fmt.Sprintf(`"%d"`, k))
+	}
+	sort.Strings(l)
+	return strings.Join(l, ",")
+}
+
+// extractStringFromUint64Slice helps extract string info from uint64 slice.
+func extractStringFromUint64Slice(slice []uint64) string {
```
Comment (IcePigZDB) [inline planner/core/util.go:311]: ditto

File: executor/memtable_reader_test.go [hunk 10]
```diff
@@ -941,3 +944,706 @@ func (s *testMemTableReaderSuite) TestTiDBClusterLogError(c *C) {
...
+
+type testTikvRegionStatusTableSuite struct {
+	*testInfoschemaTableSuiteBase
+	httpServers []*httptest.Server
+	startTime   time.Time
+}
+
+func (s *testTikvRegionStatusTableSuite) SetUpSuite(c *C) {
+	s.testInfoschemaTableSuiteBase.SetUpSuite(c)
+	store := &mockStoreWithMultiPD{
```
Comment (IcePigZDB) [inline executor/memtable_reader_test.go:1426]: how about use `mockStore` in `infoschem_reade_test.go`, because there is only one pd server here.

**Comment by nolouch** [comment]
@qidi1 please fix the conflicts.

**Pull request closed without merging**